test/src/module/common/walFilterTest.c

```diff
@@ -0,0 +1,1848 @@
+    {
+        Buffer *wal2;
+        StringList *argBaseList = strLstNew();
+        hrnCfgArgRawZ(argBaseList, cfgOptFork, CFGOPTVAL_FORK_GPDB_Z);
+        hrnCfgArgRawZ(argBaseList, cfgOptPgPath, TEST_PATH "/pg");
+        hrnCfgArgRawZ(argBaseList, cfgOptRepoPath, TEST_PATH "/repo");
+        hrnCfgArgRawZ(argBaseList, cfgOptStanza, "test1");
+        HRN_CFG_LOAD(cfgCmdArchiveGet, argBaseList);
+
+        HRN_PG_CONTROL_OVERRIDE_VERSION_PUT(
+            storagePgWrite(), PG_VERSION_94, 9420600, .systemId = HRN_PG_SYSTEMID_94, .catalogVersion = 301908232,
+            .pageSize = DEFAULT_GDPB_XLOG_PAGE_SIZE, .walSegmentSize = 64 * 1024 * 1024);
+
+        const PgControl pgControl = {
+            .version = PG_VERSION_94,
+            .pageSize = DEFAULT_GDPB_PAGE_SIZE,
+            .walPageSize = DEFAULT_GDPB_XLOG_PAGE_SIZE,
+            .walSegmentSize = 64 * 1024 * 1024
+        };
+
+        HRN_INFO_PUT(
+            storageRepoWrite(), INFO_ARCHIVE_PATH_FILE,
+            "[db]\n"
+            "db-id=1\n"
+            "db-system-id=7374327172765320188\n"
+            "db-version=\"9.4\"\n"
+            "\n"
+            "[db:history]\n"
+            "1={\"db-id\":" HRN_PG_SYSTEMID_94_Z ",\"db-version\":\"9.4\"}");
+
+        HRN_STORAGE_PATH_CREATE(storageRepoIdxWrite(0), STORAGE_REPO_ARCHIVE "/9.4-1");
+        ArchiveGetFile archiveInfo = {
+            .file = STRDEF(
+                STORAGE_REPO_ARCHIVE "/9.4-1/0000000100000000/000000010000000000000002-abcdabcdabcdabcdabcdabcdabcdabcdabcdabcd.gz"),
+            .repoIdx = 0,
+            .archiveId = STRDEF("9.4-1"),
+            .cipherType = cipherTypeNone,
+            .cipherPassArchive = STRDEF("")
+        };
+
+        TEST_TITLE("simple read begin from prev file");
+        filter = walFilterNew(CFGOPTVAL_FORK_GPDB, pgControl, &archiveInfo);
+        {
+            Buffer *wal1 = bufNew(32768);
+
+            record = hrnGpdbCreateXRecord(RM_XLOG_ID, XLOG_NOOP, 32688, NULL);
+            hrnGpdbWalInsertXRecordSimple(wal1, record);
+
+            record = hrnGpdbCreateXRecord(RM_XLOG_ID, XLOG_NOOP, 100, NULL);
+            hrnGpdbWalInsertXRecordP(wal1, record, INCOMPLETE_RECORD);
+            fillLastPage(wal1, DEFAULT_GDPB_XLOG_PAGE_SIZE);
+
+            HRN_STORAGE_PUT(
+                storageRepoWrite(),
+                STORAGE_REPO_ARCHIVE "/9.4-1/0000000100000000/000000010000000000000001-abcdabcdabcdabcdabcdabcdabcdabcdabcdabcd",
+                wal1, .compressType = compressTypeGz);
+        }
+
+        wal2 = bufNew(1024 * 1024);
+        record = hrnGpdbCreateXRecord(RM_XLOG_ID, XLOG_NOOP, 100, NULL);
+        hrnGpdbWalInsertXRecordP(wal2, record, NO_FLAGS, .segno = 2, .begin_offset = 132 - 8);
+        record = hrnGpdbCreateXRecord(0, XLOG_SWITCH, 0, NULL);
+        hrnGpdbWalInsertXRecordSimple(wal2, record);
+
+        fillLastPage(wal2, DEFAULT_GDPB_XLOG_PAGE_SIZE);
+        result = testFilter(filter, wal2, bufSize(wal2), bufSize(wal2));
```

Comment:
`testFilter` allocates memory for the `result`, and `result = testFilter(...)` is called multiple times. But result is never freed. Why not to add ``bufFree(result)` after each test?

Reply:
I wrapped each test in a temporary context so that the memory will be freed after each test.